### PR TITLE
feat(gnokey): Print out the transaction hash when maketx executes successfully

### PR DESCRIPTION
**Pull request by linhpn99**
Relate to https://github.com/gnolang/gno/issues/2303

<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

File: tm2/pkg/crypto/keys/client/broadcast.go [hunk 1]
```diff
@@ -87,6 +88,7 @@ func execBroadcast(cfg *BroadcastCfg, args []string, io commands.IO) error {
 		io.Println("GAS USED:  ", res.DeliverTx.GasUsed)
 		io.Println("HEIGHT:    ", res.Height)
 		io.Println("EVENTS:    ", string(res.DeliverTx.EncodeEvents()))
+		io.Println("TX HASH:   ", hex.EncodeToString(res.Hash))
```
Comment (zivkovicmilos) [inline tm2/pkg/crypto/keys/client/broadcast.go:91]: We have an "unofficial standard" on how we generate and display tx hashes:

The transaction hash (which is the sha'd amino binary encoding) is **base64 encoded**, and displayed as such throughout all of our tools (ex. Gnoscan, tx-indexer...). We also use this format for the `tx` RPC endpoint

So here, you'd need to base64 it

File: tm2/pkg/crypto/keys/client/maketx.go [hunk 2]
```diff
@@ -219,6 +220,7 @@ func ExecSignAndBroadcast(
 	io.Println("GAS USED:  ", bres.DeliverTx.GasUsed)
 	io.Println("HEIGHT:    ", bres.Height)
 	io.Println("EVENTS:    ", string(bres.DeliverTx.EncodeEvents()))
+	io.Println("TX HASH:   ", hex.EncodeToString(bres.Hash))
```
Comment (zivkovicmilos) [inline tm2/pkg/crypto/keys/client/maketx.go:223]: Same as the previous comment about base64 encoding 🙏

File: tm2/pkg/crypto/keys/client/broadcast.go (hunk 1 above)
Comment (zivkovicmilos) [reply tm2/pkg/crypto/keys/client/broadcast.go:91]: It is worth noting a quirk, we autodecode base64 into bytes on the RPC layer, but the index we save for the transactions is the actual transaction hash in bytes :)

Doesn't change anything in the comment above, just gives additional context

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
Thank you for the addition 🙏 

We just need to change the display encoding, and we should be good to go 🚀

**Comment by zivkovicmilos** [comment]
@linhpn99 can you please merge in the master branch? 🙏

File: tm2/pkg/crypto/keys/client/broadcast.go (hunk 1 above)
Comment (linhpn99) [reply tm2/pkg/crypto/keys/client/broadcast.go]: Thanks for your info, I will use base64 instead

**Comment by moul** [comment]
Can we have at least one unit test not checking that "a string is printed", but actually checking that the hash looks correct?

**Comment by linhpn99** [comment]
> Can we have at least one unit test not checking that "a string is printed", but actually checking that the hash looks correct?

for sure, i'm working on it

**Comment by linhpn99** [comment]
> Can we have at least one unit test not checking that "a string is printed", but actually checking that the hash looks correct?

However, with the current code structure, it is not possible to write unit tests due to the need to mock certain components. I think it would be better to create a separate PR to address this issue, which would include the necessary mocking and test cases for the missing functions. wdyt @moul @zivkovicmilos ?

File: tm2/pkg/crypto/keys/client/broadcast.go (hunk 1 above)
Comment (linhpn99) [reply tm2/pkg/crypto/keys/client/broadcast.go]: https://github.com/gnolang/gno/pull/2309/commits/c1eb1fe7f6ce25dc3389b905e79b3799312e2c86

File: tm2/pkg/crypto/keys/client/maketx.go (hunk 2 above)
Comment (linhpn99) [reply tm2/pkg/crypto/keys/client/maketx.go]: https://github.com/gnolang/gno/pull/2309/commits/c1eb1fe7f6ce25dc3389b905e79b3799312e2c86

File: tm2/pkg/crypto/keys/client/maketx.go [hunk 3]
```diff
@@ -219,6 +221,7 @@ func ExecSignAndBroadcast(
 	io.Println("GAS USED:  ", bres.DeliverTx.GasUsed)
 	io.Println("HEIGHT:    ", bres.Height)
 	io.Println("EVENTS:    ", string(bres.DeliverTx.EncodeEvents()))
+	io.Println("TX HASH:   ", url.QueryEscape(base64.StdEncoding.EncodeToString(bres.Hash)))
```
Comment (linhpn99) [inline tm2/pkg/crypto/keys/client/maketx.go:224]: @zivkovicmilos During testing, I observed that when the hash returned in base64 format contains special characters like this one, `8y9J740MXZUhlgUv+Kxt+yVRqtTuFjHpNdLZPkZVhR4=`, the hash parsed from the HTTP server side converts special characters into spaces. This function will help transform special characters into their corresponding base64 string `8y9J740MXZUhlgUv%2BKxt%2ByVRqtTuFjHpNdLZPkZVhR4%3D` while preserving the integrity of the original string :

File: tm2/pkg/crypto/keys/client/broadcast.go [hunk 4]
```diff
@@ -87,6 +89,7 @@ func execBroadcast(cfg *BroadcastCfg, args []string, io commands.IO) error {
 		io.Println("GAS USED:  ", res.DeliverTx.GasUsed)
 		io.Println("HEIGHT:    ", res.Height)
 		io.Println("EVENTS:    ", string(res.DeliverTx.EncodeEvents()))
+		io.Println("TX HASH:   ", url.QueryEscape(base64.StdEncoding.EncodeToString(res.Hash)))
```
Comment (linhpn99) [inline tm2/pkg/crypto/keys/client/broadcast.go:92]: @zivkovicmilos During testing, I observed that when the hash returned in base64 format contains special characters like this one, `8y9J740MXZUhlgUv+Kxt+yVRqtTuFjHpNdLZPkZVhR4=`, the hash parsed from the HTTP server side converts special characters into spaces. This function will help transform special characters into their corresponding base64 string `8y9J740MXZUhlgUv%2BKxt%2ByVRqtTuFjHpNdLZPkZVhR4%3D` while preserving the integrity of the original string

File: tm2/pkg/crypto/keys/client/broadcast.go [hunk 5]
```diff
@@ -87,6 +88,7 @@ func execBroadcast(cfg *BroadcastCfg, args []string, io commands.IO) error {
 		io.Println("GAS USED:  ", res.DeliverTx.GasUsed)
 		io.Println("HEIGHT:    ", res.Height)
 		io.Println("EVENTS:    ", string(res.DeliverTx.EncodeEvents()))
+		io.Println("TX HASH:   ", base64.StdEncoding.EncodeToString(res.Hash))
```
Comment (zivkovicmilos) [inline tm2/pkg/crypto/keys/client/broadcast.go:91]: This should be also printed in the other case, when the transaction's delivery failed (it's a good transaction, but failed when executing):

`if res.DeliverTx.IsErr() {...}`

File: tm2/pkg/crypto/keys/client/maketx.go [hunk 6]
```diff
@@ -219,6 +220,7 @@ func ExecSignAndBroadcast(
 	io.Println("GAS USED:  ", bres.DeliverTx.GasUsed)
 	io.Println("HEIGHT:    ", bres.Height)
 	io.Println("EVENTS:    ", string(bres.DeliverTx.EncodeEvents()))
+	io.Println("TX HASH:   ", base64.StdEncoding.EncodeToString(bres.Hash))
```
Comment (zivkovicmilos) [inline tm2/pkg/crypto/keys/client/maketx.go:223]: Same comment as before, should also be printed in the other case when the `DeliverTx.IsErr() == true`

**Review by zivkovicmilos — APPROVED** [review]
Looks good now 💯 

I've left a small comment where we can additionally print the hash, otherwise we're good to go 🚀

File: tm2/pkg/crypto/keys/client/maketx.go (hunk 6 above)
Comment (linhpn99) [reply tm2/pkg/crypto/keys/client/maketx.go:224]: https://github.com/gnolang/gno/pull/2309/commits/9c3a0f694871ffa3625387ae10ac410d6bfebee0

File: tm2/pkg/crypto/keys/client/broadcast.go (hunk 5 above)
Comment (linhpn99) [reply tm2/pkg/crypto/keys/client/broadcast.go:92]: https://github.com/gnolang/gno/pull/2309/commits/9c3a0f694871ffa3625387ae10ac410d6bfebee0

**Pull request merged**